### PR TITLE
Move findTraces() to a backend service

### PR DESCRIPTION
**Pull request by PatrickTasse**
Signed-off-by: Patrick Tasse <patrick.tasse@ericsson.com>

### CODE REVIEW

File: theia-extensions/viewer-prototype/src/common/backend-service.ts [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import { CancellationToken } from '@theia/core';
+
+export const backendServicePath = '/services/backend-service';
+export const BackendService = Symbol('BackendService');
+
+export interface BackendService {
```
Comment (bhufmann) [inline theia-extensions/viewer-prototype/src/common/backend-service.ts:6]: Should the name be a bit clearer? It feels that name `BackendService` is too general

File: theia-extensions/viewer-prototype/src/node/backend-service-impl.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
+import { injectable } from 'inversify';
+import fs = require('fs');
+import { Dirent } from 'fs';
+import { Path } from '@theia/core/lib/common/path';
+import { CancellationToken } from '@theia/core';
+import { BackendService } from '../common/backend-service';
+
+@injectable()
+export class BackendServiceImpl implements BackendService {
```
Comment (bhufmann) [inline theia-extensions/viewer-prototype/src/node/backend-service-impl.ts:9]: Same comment about name.

File: theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer.tsx [hunk 3]
```diff
@@ -136,36 +136,22 @@ export class TraceViewerWidget extends ReactWidget {
...
-                    if (fileStat) {
-                        if (fileStat.isDirectory) {
-                            // Find recursivly CTF traces
-                            await this.findTraces(fileStat, tracesArray, isCancelled);
-                        } else {
-                            // Open single trace file
-                            tracesArray.push(this.uri);
-                        }
-                    }
+                    const tracesArray = await this.backendService.findTraces(this.uri.toString(), cancellation.token);
```
Comment (bhufmann) [inline theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer.tsx:154]: I pressed cancel while having a breakpoint in the backend service to test the cancellation. In the backend it stopped, but it returned an empty `tracesArray`. So, the check below is then true and I get the error message, `No valid traces found in selected directory ...`. When the array length is 0 and the cancellation was requested I think the error message should not be shown.

**Review by bhufmann — CHANGES_REQUESTED** [review]
Thanks for the contribution

File: theia-extensions/viewer-prototype/src/common/backend-service.ts [hunk 4]
```diff
@@ -0,0 +1,8 @@
+import { CancellationToken } from '@theia/core';
+
+export const backendServicePath = '/services/backend-service';
```
Comment (bhufmann) [inline theia-extensions/viewer-prototype/src/common/backend-service.ts:3]: Can there be a name conflict with other extensions? Make it needs to be more specific.

File: theia-extensions/viewer-prototype/src/common/backend-service.ts (hunk 1 above)
Comment (PatrickTasse) [reply theia-extensions/viewer-prototype/src/common/backend-service.ts]: Changed to BackendFileService

File: theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer.tsx (hunk 3 above)
Comment (PatrickTasse) [reply theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer.tsx]: Moved cancellation check before array size check.

File: theia-extensions/viewer-prototype/src/common/backend-service.ts (hunk 4 above)
Comment (PatrickTasse) [reply theia-extensions/viewer-prototype/src/common/backend-service.ts]: Changed prefix to /services/theia-trace-extension/<service name>, for the existing service also.

**Review by bhufmann — DISMISSED** [review]
Looks good to me. Thanks!

File: theia-extensions/viewer-prototype/src/node/backend-file-service-impl.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        } else if (stats.isFile()) {
+            traces.push(path);
+        }
+        if (cancellationToken.isCancellationRequested) {
+            return [];
+        }
+        return traces;
+    }
+
+    protected async findCtfTraces(path: string, traces: string[], cancellationToken: CancellationToken): Promise<void> {
```
Comment (MatthewKhouzam) [inline theia-extensions/viewer-prototype/src/node/backend-file-service-impl.ts:25]: Can we have a findDirTrace... uftrace won't work with this and findTraces.

**Review by MatthewKhouzam — CHANGES_REQUESTED** [review]
I can remove the "request change" if you think uftrace is not a priority.

File: theia-extensions/viewer-prototype/src/node/backend-file-service-impl.ts (hunk 5 above)
Comment (PatrickTasse) [reply theia-extensions/viewer-prototype/src/node/backend-file-service-impl.ts:25]: I renamed it to deepFindTraces() and added Javadoc to specify that it currently only finds CTF traces.

**Review by MatthewKhouzam — APPROVED** [review]
In the words of the kool aid man...
Oh Yeah!

**Review by bhufmann — APPROVED** [review]
Thanks!

**Pull request merged**